app/master/build_request_handler.py

```diff
@@ -101,6 +101,11 @@ def _prepare_build_async(self, build, project_lock):
                                    log_msg='Build preparation loop is handling request for build {build_id}.')
             try:
                 build.prepare(self._subjob_calculator)
+                # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                if len(build.all_subjobs()) == 0:
+                    self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                    build.mark_started()
+                    build.perform_async_postbuild_tasks()
```

Comment:
Another option, more along the lines of what you had originally, would be inside the build's `finish()` method to just progress to the BUILDING state first if the build is in the PREPARED state. I'm not opposed to that either.